### PR TITLE
KAFKA-8927: Deprecate PartitionGrouper interface

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @vvcephei @ableegoldman @abbccdda 

Note: Don't merge before the KIP is accepted.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:40]: We cannot suppress deprecation warnings for imports -- hence, we need to remove imports of deprecated classed and use the fully qualified name in the code (applied to other classes in this PR, too).

**Review by guozhangwang — APPROVED** [review]
The change LGTM.

**Comment by mjsax** [comment]
Java 11 / 2.12 passed. Other two failed (results are not longer available).

Retest this please.

Call for second review @cadonna @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java]: Since we have a Jira ticket that is even referenced here, I would prefer to remove the ToDo from the code.

Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java]: Since we have a Jira ticket that is even referenced here, I would prefer to remove the ToDo from the code.

File: streams/src/main/java/org/apache/kafka/streams/processor/PartitionGrouper.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/PartitionGrouper.java]: Since we have a Jira ticket that is even referenced here, I would prefer to remove the ToDo from the code.

**Review by cadonna — APPROVED** [review]
LGTM, Thank you @mjsax 

Just some nits

**Comment by mjsax** [comment]
Java 11 / 2.12 passed. The other two failed with env issues.

Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java]: What is bad about the `TODO` ?

**Comment by guozhangwang** [comment]
ping @bbejeck for reviewing and finally merging this PR :)

File: streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java]: `TODO` pollute the code IMO. They tend to be forgotten or they describe something that is obsolete or even worse the description is not comprehensible. For all of these reasons, I try to avoid `TODO`s as much as I can. Using Jira tickets to document `TODO`s is much more effective and does not pollute code. Since we have already a Jira ticket that tells us what is needed to be done in this case, having a TODO is redundant. Up to you, whether you want to follow my reasoning or not. I already approved this PR.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java]: Ack.

**Comment by mjsax** [comment]
Java 11 / 2.12 passed.

Java 11 / 2.13 and Java 8 failed. Test results not longer available.

Retest this please.

**Pull request merged**
